proxy/server/target.go

```diff
@@ -211,13 +235,37 @@ func (s *TargetStream) Run(nonce uint32, replyChan chan *pb.ProxyReply) {
 					// reqChan was closed, meaning no more messages are incoming
 					// if this was a client stream, we issue a half-close
 					if s.serviceMethod.ClientStreams() {
-						return grpcStream.CloseSend()
+						err = grpcStream.CloseSend()
+						if err != nil {
+							s.cancelFunc()
+						}
+						return err
 					}
 					// otherwise, we're done
 					return nil
 				}
 			}
-			err := grpcStream.SendMsg(req)
+
+			authinput, err := rpcauth.NewRPCAuthInput(ctx, s.Method(), req)
+			if err != nil {
+				err = status.Errorf(codes.Internal, "error creating authz input %v", err)
+				s.errChan <- err
```

Comment:
you can use 'CloseWith' here,  does the same thing, but handles the non-blocking send to errChan and call to cancel(). This prevents the case in which you happen to hit multiple errors at the same time, yielding a deadlock in sending to errChan

Reply:
or alternately, you could just return the error? errGroup cancels the ctx on the first non-err return, and should return the first error encountered from Wait(), so I'm not sure you need to write to errChan directly. Err chan is mostly for 'external' cancellation from outside the stream.